tools/depscan/main.go

```diff
@@ -0,0 +1,302 @@
+		Use: "scan",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return RunLicenseScan(cmd.Context(), opts)
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.Binary, "binary", opts.Binary, "binary to analyze")
+
+	cmd.Flags().BoolVar(&opts.IncludeLicenses, "print", opts.IncludeLicenses, "include license text")
+
+	cmd.Flags().StringArrayVar(&opts.IgnorePackage, "ignore", opts.IgnorePackage, "packages to ignore")
+
+	return cmd
+}
+
+type RunLicenseScanOptions struct {
+	Binary string
+
+	IgnorePackage []string
+
+	IncludeLicenses bool
+}
+
+func RunLicenseScan(ctx context.Context, opts RunLicenseScanOptions) error {
+	buildInfo, err := buildinfo.ReadFile(opts.Binary)
+	if err != nil {
+		return fmt.Errorf("error reading binary info from %q: %w", opts.Binary, err)
+	}
+
+	var errors []error
+
+	var modules []*Module
+	for _, dep := range buildInfo.Deps {
+		for _, ignorePackage := range opts.IgnorePackage {
+			if ignorePackage == dep.Path {
+				klog.Infof("ignoring package %s@%s", dep.Path, dep.Version)
+				continue
+			}
+		}
+
+		module := &Module{
+			// Path: dep.Path,
+			Name:    dep.Path,
+			Sum:     dep.Sum,
+			Version: dep.Version,
+		}
+		if dep.Replace != nil {
+			module.Version = dep.Replace.Version
+		}
+		modules = append(modules, module)
+
+		p := filepath.Join("modules", module.Name, module.Version+".yaml")
+		b, err := ioutil.ReadFile(p)
+		if err != nil {
+			if os.IsNotExist(err) {
+				if err := os.MkdirAll(filepath.Dir(p), 0755); err != nil {
+					return fmt.Errorf("error from Mkdir(%q): %w", filepath.Dir(p), err)
+				}
+				s := "# https://pkg.go.dev/" + module.Name + "@" + module.Version + "\n"
+				s += "license: TODO"
+				b = []byte(s)
+				if err := ioutil.WriteFile(p, b, 0644); err != nil {
+					return fmt.Errorf("error writing %q: %w", p, err)
+				}
+			} else {
+				return fmt.Errorf("error reading %q: %w", p, err)
+			}
+		}
+		info := &moduleInfo{}
+		if err := yaml.Unmarshal(b, info); err != nil {
+			return fmt.Errorf("error parsing %q: %w", p, err)
+		}
+		info.License = strings.TrimSpace(info.License)
+		if info.License == "TODO" || info.License == "" {
+			errors = append(errors, fmt.Errorf("license not known for %s@%s", module.Name, module.Version))
+		}
+		module.Info = info
+		//klog.Infof("module %v: %#v", module, info)
+	}
+
+	if len(errors) == 0 && opts.IncludeLicenses {
+		// binaryName := filepath.Base(opts.Binary)
+		for _, module := range modules {
+			if licenseFiles, err := includeLicense(ctx, module); err != nil {
+				errors = append(errors, fmt.Errorf("error getting license for %s@%s: %w", module.Name, module.Version, err))
+			} else {
+				module.LicenseFiles = licenseFiles
+			}
+		}
+	}
+
+	mustShipCodeByLicense := map[string]bool{
+		"APACHE-2.0":       false,
+		"BSD-3-CLAUSE":     false,
+		"BSD-2-CLAUSE":     false,
+		"MIT":              false,
+		"UNICODE-DFS-2016": false,
+		"MPL-2.0":          true,
+		"PUBLICDOMAIN":     false,
+		"ISC":              false,
+	}
```

Comment:
you forgot GPL :)

Reply:
I added libraries that we're using in various binaries where I've tried this out ... no GPL yet!